File: Hard/312. Burst Balloons/BurstBalloons.py
'''
LeetCode Hard: 312. Burst Balloons
https://leetcode.com/problems/burst-balloons/

Check BurstBalloons.cpp for brief solution explanation.
'''
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        N = len(nums)
        nums.insert(N, 1)
        nums.insert(0, 1)
        dp = [[0 for i in range(N+1)] for i in range(N+1)]
        for sz in range(1, N+1):
            for l in range(1, N-sz+2):
                r = l+sz-1
                for m in range(l, r+1):
                    res = nums[l-1]*nums[m]*nums[r+1]
                    if l<=m-1:
                        res+=dp[l][m-1]
                    if m+1<=r:
                        res+=dp[m+1][r]
                    dp[l][r] = max(dp[l][r], res)
        return dp[1][N]

File: Hard/312. Burst Balloons/BurstBalloons.java
/*
LeetCode Hard: 312. Burst Balloons
https://leetcode.com/problems/burst-balloons/

Check BurstBalloons.cpp for brief solution explanation.
*/
class Solution {
    public int maxCoins(int[] nums) {
        int N = nums.length;
        ArrayList<Integer>vals = new ArrayList<Integer>();
        vals.add(1);
        for(int i:nums){
            vals.add(i);
        
        vals.add(1);
        int[][]dp = new int[N+1][N+1];
        for(int len = 1;len<=N;len++){
            for(int l = 1;l<=N-len+1;l++){
                int r = l+len-1;
                for(int m = l;m<=r;m++){
                    int res = vals.get(l-1)*vals.get(m)*vals.get(r+1);
                    if(l<=m-1){
                        res+=dp[l][m-1];
                    }
                    if(m+1<=r){
                        res+=dp[m+1][r];
                    }
                    dp[l][r] = Math.max(dp[l][r], res);
                }
            }
        }
        return dp[1][N];
    }
}